flax/errors.py

```diff
@@ -297,5 +323,263 @@ def __call__(self, x):
+  """
+  You can only assign Module attributes to ``self`` inside
+  :meth:`Module.setup() <flax.linen.Module.setup>`. Outside of that method, the
+  Module instance is frozen (i.e., immutable). This behavior is similar to
+  frozen Python dataclasses.
+  
+  For instance, this error is raised in the following case::
+
+    class SomeModule(nn.Module):
+      @nn.compact
+      def __call__(self, x, num_features=10):
+        self.num_features = num_features  # <-- ERROR!
+        x = nn.Dense(self.num_features)(x)
+        return x
+
+    s = SomeModule().init(random.PRNGKey(0), jnp.ones((5, 5)))
+
+  Similarly, the error is raised when trying to modify a submodule's attributes
+  after constructing it, even if this is done in the ``setup()`` method of the
+  parent module::
+
+    class Foo(nn.Module):
+        def setup(self):
+          self.dense = nn.Dense(features=10)
+          self.dense.features = 20  # <--- This is not allowed
+        
+        def __call__(self, x):
+          return self.dense(x)
+  """
+  def __init__(self, module_cls, attr_name, attr_val):
+    super().__init__(f'Can\'t set {attr_name}={attr_val} for Module of type '
+                    f'{module_cls}: Module instance is frozen outside of '
+                     'setup method.')
+
+
+class MultipleMethodsCompactError(FlaxError):
+  """
+  The ``@compact`` decorator may only be added to at most one method in a Flax
+  module. In order to resolve this, you can:
+  
+  * remove ``@compact`` and define submodules and variables using 
+    :meth:`Module.setup() <flax.linen.Module.setup>`.
+  * Use two separate modules that both have a unique ``@compact`` method.
+  """
+  def __init__(self):
+    super().__init__(f'Only one method per class can be @compact')
+
+class ReservedModuleAttributeError(FlaxError):
+  """
+  This error is thrown when creating a Module that is using reserved attributes.
+  The following attributes are reserved:
+  
+  * ``parent``: The parent Module of this Module.
+  * ``name``: The name of this Module.
+  """
+  def __init__(self, annotations):
+    super().__init__(f'properties `parent` and `name` are reserved: '
+                     f'{annotations}')
+
+
+class InitModuleInvalidRngsError(FlaxError):
+  """
+  This error is thrown if the input rngs to 
+  :meth:`Module.init() <flax.linen.Module.init>` are incorrect.
+  See :class:`InvalidRngError` for more details.
+  """
+  def __init__(self, module_name, rngs):
+    super().__init__(f'RNGs should be of shape (2,) in Module {module_name}, '
+                     f'but rngs are: {rngs}')
+
+
+class ApplyModuleInvalidMethodError(FlaxError):
+  """
+  When calling :meth:`Module.apply() <flax.linen.Module.apply>`, you can specify
+  the method to apply using parameter ``method``. This error is thrown if the
+  provided parameter is not a method in the Module and not a function with at
+  least one argument.
+
+  A typical call looks as follows::
```

Comment:
I think from here on should be part of the `apply` docstring instead, and then replace this part with, e.g. "learn more on the on the reference docs for apply".